src/NLog.Mongo/MongoTarget.cs

```diff
@@ -216,38 +253,80 @@ private void AddDefaults(BsonDocument document, LogEventInfo logEvent)
 
             if (logEvent.Exception != null)
                 document.Add("Exception", CreateException(logEvent.Exception));
-
-
         }
 
         private void AddProperties(BsonDocument document, LogEventInfo logEvent)
         {
-            var propertiesDocument = new BsonDocument();
-            foreach (var field in Properties)
-            {
-                string key = field.Name;
-                var value = GetValue(field, logEvent);
-
-                if (value != null)
-                    propertiesDocument[key] = value;
-            }
-
-            var properties = logEvent.Properties ?? Enumerable.Empty<KeyValuePair<object, object>>();
-            foreach (var property in properties)
+            if (logEvent.HasProperties || Properties.Count > 0)
             {
-                if (property.Key == null || property.Value == null)
-                    continue;
+                var propertiesDocument = new BsonDocument();
+                for (int i = 0; i < Properties.Count; ++i)
+                {
+                    string key = Properties[i].Name;
+                    var value = GetValue(Properties[i], logEvent);
 
-                string key = Convert.ToString(property.Key, CultureInfo.InvariantCulture);
-                string value = Convert.ToString(property.Value, CultureInfo.InvariantCulture);
+                    if (value != null)
+                        propertiesDocument[key] = value;
+                }
 
-                if (!string.IsNullOrEmpty(value))
-                    propertiesDocument[key] = new BsonString(value);
+                if (logEvent.HasProperties)
+                {
+                    foreach (var property in logEvent.Properties)
+                    {
+                        if (property.Key == null || property.Value == null)
+                            continue;
+
+                        string key = Convert.ToString(property.Key, CultureInfo.InvariantCulture);
+                        if (string.IsNullOrEmpty(key))
+                            continue;
+
+                        if (key.IndexOf('.') >= 0)
+                            key = key.Replace('.', '_');
+
+                        TypeCode typeCode = Convert.GetTypeCode(property.Value);
+                        BsonValue bsonValue = null;
+                        switch (typeCode)
```

Comment:
I'm concerned this will be a breaking change.   All previous documents were string values.  Maybe good enough to document and make peeps aware.

Reply:
Removed breaking change. Just thought it would be nice.